Perform json-blif conversion in one step in conv.tcl

olofk opened the pull request:
Instead of calling python from the bash script wrapper, call
it directly from conv.tcl

Signed-off-by: Olof Kindgren <olof.kindgren@gmail.com>

olofk commented:
Forgot to mention, this is kind of a first step in my objective to eventually let Edalize call the individual EDA tools directly instead of going through the shell script wrappers.

olofk commented:
Replaces https://github.com/SymbiFlow/symbiflow-arch-defs/pull/1461

olofk commented:
Ok, I redid the PR. Updating topic.

The new PR moves most of the code out of synth.tcl and conv.tcl into functions, so that these functions can be called by something else (e.g. Edalize) directly instead of going through symbiflow_synth. This means that the code is still backwards-compatible.

A few notes:
1. The code from synth.tcl moved into new file synth_functions.tcl. Perhaps extending utils.tcl would make more sense?
2. I tried to create functions in synth_functions.tcl that grouped together commands into what I believed was a full function. This might be very wrong though as I don't fully understand all the logic in synth.tcl. I could move stuff around or simply have one big function that does all of synth.tcl
3. I reduced the number of env vars as much as possible and only made $SHARE_DIR_PATH and $TOP required. The other env vars were effectively hard coded anyway. I did keep USE_ROI and USE_LUT_CONSTANTS though as those felt more like global debug flags. But again, I'm not sure I got that right

acomodi commented on xc/xc7/toolchain_wrappers/symbiflow_synth:
```diff
@@ -5,13 +5,9 @@ MYPATH=`realpath $0`
 MYPATH=`dirname ${MYPATH}`
 
 export SHARE_DIR_PATH=`realpath ${MYPATH}/../share/symbiflow`
-export TECHMAP_PATH=${SHARE_DIR_PATH}/techmaps/xc7_vpr/techmap
```
The `TECHMAP_PATH` may change based on where the synth is executed. For instance, in the `symbiflow-arch-defs` tests, the techmap path corresponds to this `https://github.com/SymbiFlow/symbiflow-arch-defs/tree/master/xc/xc7/techmap`, while in the installed toolchain, the techmap PATH is in a different location as specified in here: https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/xc/common/cmake/install.cmake#L74.

I am not quite sure how to handle different paths locations without having the `TECHMAP_PATH` variable

olofk replied:
I see. Let's leave TECHMAP_PATH in there then. I'll fix

acomodi reviewed (changes requested):
Hi @olofk, thanks for this PR! I believe that making the synth scripts lighter and extracting the various steps in different processes is quite a good enhancement.

I left a few comments, other than that it looks good

olofk commented on xc/xc7/yosys/synth.tcl:
```diff
@@ -64,121 +39,16 @@ update_pll_and_mmcm_params
-clean_processes
-write_json $::env(OUT_JSON).carry_fixup.json
-
-exec $::env(PYTHON3) $::env(UTILS_PATH)/fix_xc7_carry.py < $::env(OUT_JSON).carry_fixup.json > $::env(OUT_JSON).carry_fixup_out.json
```
Got it. I'll change to use the env var if it exists and call python3 directly. Sounds good?

acomodi replied:
Yeah, sounds good

olofk commented:
Updated now with restored TECHMAP_PATH and only calling python3 directly if there is no PYTHON3 env var

One thing. Do we need to register the new synth_functions.tcl in some CMake script or so to make sure it is installed? And do we want to move the synth_functions contents into utils.tcl instead or is it fine to have a new file for this?

acomodi commented:
@olofk Good question, I missed this one actually. So IMO it might be ok to have all the synth functions alongside with the ones in the already present `utils.tcl` file, which is already getting installed along with the other scripts.

olofk commented:
I added all the functions to utils.tcl and updated the PR. It feels like they are similar to the ones already found there so I agree it make sense to have them all there

olofk commented:
I see that CI failed. Is this the issue? `[TCL: yosys -import] Command name collision: found pre-existing command `wbfliERROR: TCL interpreter returned an error: can't read "::env(TOP)": no such variable`

Think it might be. Force-pushing with this change
```
diff --git a/xc/xc7/yosys/synth.tcl b/xc/xc7/yosys/synth.tcl
index 55592d92..66ab227d 100644
--- a/xc/xc7/yosys/synth.tcl
+++ b/xc/xc7/yosys/synth.tcl
@@ -11,7 +11,7 @@ yosys -import
 
 source [file join [file normalize [info script]] .. utils.tcl]
 
-if { [info exists $::env(TOP)]} {
+if { [info exists ::env(TOP)]} {
     set top $::env(TOP)
 } else {
     set top ""
```

acomodi reviewed (commented):
So, looking at this one more time, I think that we'd need a way to keep the old behavior of allowing overriding file names with env variable. Mainly because this way we can keep what is generated during synthesis with what is specified in the CMake infrastructure (https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/common/cmake/devices.cmake#L1499-L1616).

For instance, renaming `$::env(OUT_SDC)` to `$::env(TOP).sdc` breaks this sync, allowing for situations in which the makefiles generated have targets that could fail, as the naming of the generated files and what was generated by CMake do not coincide.

I believe that a possible way to reduce duplication is adding a util function that, given a name of an env variable and a default name as inputs, checks if the env variable exists and returns its content or otherwise falls back to the default, giving the precedence to the env variable (similarly to what was done for python3).

olofk commented:
I could also restore the env vars in synth.tcl instead. As long as we can have the functions in utils.tcl free from env vars that solves my original issue. That would make the diff smaller. Should I do it like that?

olofk commented:
Updated now to restore the env vars

acomodi commented:
@olofk Can you please rebase on top of master, as there were some fixes to CI? I believe we can merge once everything goes green

olofk commented:
Done. Let's hope for the CI to say yes

olofk commented:
Oh, come on! Failed again? *This* time it must work

olofk commented:
Aarrrghh!!! Seems like there's some cmake scripts that needs updating. I have no idea what I'm doing anymore

olofk commented:
Ok, fuck it. I give up with the cmake stuff and simplified the PR instead so that it doesn't touch symbiflow_synth at all. This can't possibly go wrong

olofk commented:
Yes! All green. Good to go now @acomodi ?

acomodi commented:
@olofk Great! All CIs now are green (I forgot to run kokoro as well which is still used to generate and push new device databases)

Given that there were some other merges in between, it would be good to rebase this PR on top of main and than merge

olofk commented:
@acomodi Rebased and repushed
